Changes to support private columns

gortiz opened the pull request:
Although https://github.com/apache/pinot/pull/11327 had fixed the issue where virtual/private columns were projected when executing `select *`, it seems that a regression in https://github.com/apache/pinot/pull/11457 that makes them appear again.

This PR tries to solve the issue by overriding some extra method.

gortiz commented on pinot-query-planner/src/main/java/org/apache/pinot/query/validate/Validator.java:
```diff
@@ -124,4 +129,19 @@ private static String getColumnName(SqlIdentifier identifier) {
+      SelectScope scope,
+      final boolean includeSystemVars) {
+    if (exp.getKind() == SqlKind.IDENTIFIER) {
+      if ((((SqlIdentifier) exp).names).stream().anyMatch(s -> s.startsWith("$"))) {
```
I'm not sure if this would match all the cases we want to ignore @Jackie-Jiang @xiangfu0.

This method is called once for each field in the calcite table row.

Jackie-Jiang replied:
Does it work if user explicitly select virtual column?

xiangfu0 replied:
I think so, let's add a test for this.

Jackie-Jiang commented on pinot-query-planner/src/main/java/org/apache/pinot/query/validate/Validator.java:
```diff
@@ -124,4 +129,19 @@ private static String getColumnName(SqlIdentifier identifier) {
     return names.get(names.size() - 1);
   }
+
+  protected void addToSelectList(
```
(minor)
- Add `@Override`
- Follow Pinot Style

gortiz replied:
Fixed

gortiz commented:
> Does it work if user explicitly select virtual column?

Yes. This method is only called when adding columns in `select *`. The main issue is that is not explicitly said in Calcite's javadoc. I've added a comment including the discussion in Calcite dev email list when Julian Hyde says this is not a supported feature and invite us to add a proposal (something we should do, but I guess we want to solve the issue first).

I've added a couple of test to verify the behavior and detect a possible regression in future.

walterddr commented:
just to confirm the behavior for private column is now matching v1 right? (e.g. we dont need any additional documentation to state the differences)

gortiz commented:
Yes, there are a couple of test that ensures that we do not project _internal_ columns
